src/librbd/deep_copy/ObjectCopyRequest.cc

```diff
@@ -162,7 +162,7 @@ void ObjectCopyRequest<I>::send_read() {
     ctx, util::get_image_ctx(m_src_image_ctx), io::AIO_TYPE_READ);
 
   auto req = io::ImageDispatchSpec::create_read(
-    *m_src_image_ctx, io::IMAGE_DISPATCH_LAYER_INTERNAL_START, aio_comp,
+    *m_src_image_ctx, io::IMAGE_DISPATCH_LAYER_WRITEBACK_CACHE, aio_comp,
```

Comment:
When image writeback cache is enabled, it needs to flush data in writeback cache to OSDs. But we don't invalidate the cache, so it means clean data may be cached in RWL. The original call will always the data in the source image instead of its snapshot.

Reply:
Line 141 ```io_context->read_snap(index.second);`` doesn't work when image writeback cache is enabled and caches clean data.

Reply:
I'm not sure I follow the issue but this isn't the way to solve it.

Reply:
Let me describe the issue in more detail:
The point is that the image writeback cache is enabled for the source volume. 
When it creates a snap A form a volume, it flushes IO at first. It flushes dirty data in RWL (image writeback cache) to OSDs, and then these data are marked as clean. When reading the volume, it reads from RWL at first. 
When deep copy, it needs to read data from snap A, it sets the snap id in neorados::IOContext, and then do the read. Without the image writeback cache, it sends the request to rados, and it works well. But if data is cached (clean) in RWL, it reads the data from the cache directly. The snap_id in neorados::IOContext doesn't work.  So here to read data from snap A becomes to read data from the source volume.

Reply:
But shouldn't the problem be fixed in the PWL layer? i.e. if `io_context.read_snap.value_or(CEPH_NOSNAP) != CEPH_NOSNAP)`, then don't attempt to read from the PWL cache?

Reply:
I thought about this method. The concern is whether we should handle IOContext in other interfaces, like write/flush etc. I know that snapshots can't be updated, but is it possible that IOContext is different from the iocontext in m_image_ctx?

Reply:
Yeah, it is possible that the `IOContext`'s `read_snap` is different -- specifically from this deep-copy context.